### ISSUE TITLE
Tearing on high res vnc session

### ISSUE THREAD

**Issue by blauwepiet** [novnc/noVNC#482]
Hello,

I've been testing multiple web-based vnc clients and noVNC looks very promising. It is fast and can handle 1920x1080 resolution with a decent framerate. Although I've noticed some screen tearing both horizontal and vertical. It is more prominent in firefox compared to chrome. Is there a way to reduce this tearing?

I don't know how the frames are drawn but it looks like it is done in chunks or blocks. I think the tearing is caused by the fact that the last frame is still present and a block is already drawn on this old frame. So what I thought: Isn't it possible to increase the draw block size? In this case the block will encapsulate the whole moving part of the display and will only "cut" the none moving parts where tearing isn't visible.

These are just some wild thoughts, but maybe they can be of some use. 

To recap my main question is how can I reduce the screen tearing? Increase the draw speed somehow?

Thanks in advance!

**Comment by DirectXMan12** [comment]
One of the major factors is what encoding the VNC server is using, and how big the chunks are -- noVNC tends to handle a few big chunks in png or jpeg better than lots of small chunks in RRE form, for instance.  See #431 , for instance (unfortunately, I haven't had the time to replace noVNC's zlib implementation, but, depending on your setup, that may make a difference).

**Comment by blauwepiet** [comment]
The vnc server I am using is TurboVNC and this already uses tight encoding, which is the one with the big chunks. I've verified it by looking in the rfb object in the UI. Now I was wondering if you could change the compression level in the rfb object, because I couldn't find a compression level setting?

**Comment by blauwepiet** [comment]
Nevermind, I found the compression setting, but it is still not what I want :/.. The jpeg quality trade off against the tearing. 

It's just when I use the vncviewer of Turbovnc itself it runs perfect but whenever I use it with noVNC there is tearing. But that might have something to do with an optimized protocol between the TurboVNC server and viewer, because that protocol is slightly different than the tight protocol.

**Comment by surfsara-visualization** [comment]
> One of the major factors is what encoding the VNC server is using, 
> and how big the chunks are -- noVNC tends to handle a few big 
> chunks in png or jpeg better than lots of small chunks in RRE form, 
> for instance. 

Hmmm, wouldn't tearing be just as possible when having a screen update that is sent in two big chunks, but that get applied with a canvas draw in between? From what I read about canvas it doesn't seem to provide explicit control over when it is updated on the screen, like you do have with e.g. OpenGL (by drawing to a back buffer and then flipping front-and-back buffer). Aren't you at the mercy of the browser in that case? Especially when the size and number of updates influences how things get drawn?

**Comment by kanaka** [comment]
@blauwepiet note that vncviewer in Turbovnc is a native program. Until we move noVNC to using asm.js, it's going to be slower than a native program. It's possible even vncviewer would have tearing if you made the viewport large enough (or if you heavily loaded your CPUs with other tasks).

@surfsara-visualization yes, that's true, tearing can be an artifact of the encoding process and not just the way the client is rendering. I have considered in the past adding double buffering to noVNC so that the actual drawing happens in a background canvas and this is flipped to the front after a rendering a completed frame buffer update message. However there are two issues with this:
* noVNC has really been optimized for latency and not fidelity. So I chose to render information as quickly as I can. Double buffering slows down rendering slightly in order to increase fidelity (unless you can always render the whole frame before the next screen refresh cycle in which case it's imperceptible). With the current noVNC code, large frame buffer updates may not be able to complete entirely between screen updates and so you might get the appearance of tearing. Increasing the performance of noVNC (via asm.js and binary arrays) could certainly help this and might be fast enough to enable double buffering for example.
* The VNC protocol (RFB) is incremental so depending on how the server is choosing to send frame updates, you may get the appearance of tearing no matter how fast the client is and even with double buffering. For example, consider the theoretical case where an VNC server always sends changes from the top half of the viewport and then sends the bottom half in a separate update. As the VNC server gets loaded or the network gets congestion, the client would have the appearance of tearing because the bottom half of the screen would lag the top half.

**Comment by samhed** [comment]
Fixed now seeing as #699 is merged.
